File: modules/core-api/src/main/java/com/enonic/xp/content/ExtraData.java
package com.enonic.xp.content;

import java.util.Objects;

import com.google.common.base.Preconditions;

import com.enonic.xp.data.PropertyTree;
import com.enonic.xp.module.ModuleKey;
import com.enonic.xp.schema.mixin.MixinName;

public final class ExtraData
{
    private MixinName name;

    private PropertyTree data;

    public ExtraData( final MixinName name, final PropertyTree data )
    {
        Preconditions.checkNotNull( name, "name cannot be null" );
        Preconditions.checkNotNull( data, "data cannot be null" );
        this.name = name;
        this.data = data;
    }

    public PropertyTree getData()
    {
        return data;
    }

    public void setData( final PropertyTree data )
    {
        this.data = data;
    }

    public MixinName getName()
    {
        return name;
    }

    public String getModulePrefix()
    {
        return name.getModuleKey() == null ? "" : name.getModuleKey().toString().replace( '.', '-' );
    }

    public static ModuleKey fromModulePrefix( final String modulePrefix )
    {
        return ModuleKey.from( modulePrefix.replace( '-', '.' ) );
    }

    public void setName( final MixinName name )
    {
        this.name = name;
    }

    public ExtraData copy()
    {
        return new ExtraData( name, data.copy() );
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }

        final ExtraData extraData = (ExtraData) o;

        if ( !name.equals( extraData.name ) )
        {
            return false;
        }

        if ( !data.equals( extraData.data ) )
        {
            return false;
        }

        return true;
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( name, data );
    }
}

File: modules/admin-ui/src/main/resources/web/admin/common/js/content/ExtraData.ts
module api.content {

    import MixinName = api.schema.mixin.MixinName;
    import PropertyTree = api.data.PropertyTree;
    import PropertyIdProvider = api.data.PropertyIdProvider;

    export class ExtraData implements api.Cloneable, api.Equitable {

        private name: MixinName;

        private data: PropertyTree;

        constructor(name: MixinName, data: PropertyTree) {
            this.name = name;
            this.data = data;
        }

        getName(): MixinName {
            return this.name;
        }

        getData(): PropertyTree {
            return this.data;
        }

        clone(): ExtraData {
            return new ExtraData(this.name, this.data.copy());
        }

        equals(o: api.Equitable): boolean {
            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, ExtraData)) {
                return false;
            }

            var other = <ExtraData>o;

            if (!api.ObjectHelper.equals(this.name, other.name)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.data, other.data)) {
                return false;
            }

            return true;
        }

        toJson(): api.content.json.ExtraDataJson {
            return {
                name: this.name.toString(),
                data: this.data.toJson()
            };
        }

        static fromJson(metadataJson: api.content.json.ExtraDataJson, propertyIdProvider: PropertyIdProvider): ExtraData {
            return new ExtraData(new MixinName(metadataJson.name), PropertyTree.fromJson(metadataJson.data, propertyIdProvider));
        }

        
    }

}